components/resource_control/src/future.rs

```diff
@@ -44,3 +53,251 @@ impl<F: Future> Future for ControlledFuture<F> {
         res
     }
 }
+
+#[cfg(not(test))]
+fn get_thread_io_bytes_stats() -> Result<IoBytes, String> {
+    file_system::get_thread_io_bytes_total()
+}
+
+#[cfg(test)]
+fn get_thread_io_bytes_stats() -> Result<IoBytes, String> {
+    use std::cell::Cell;
+
+    fail::fail_point!("failed_to_get_thread_io_bytes_stats", |_| {
+        Err("get_thread_io_bytes_total failed".into())
+    });
+    thread_local! {
+        static TOTAL_BYTES: Cell<IoBytes> = Cell::new(IoBytes::default());
+    }
+
+    let mut new_bytes = TOTAL_BYTES.get();
+    new_bytes.read += 100;
+    new_bytes.write += 50;
+    TOTAL_BYTES.set(new_bytes);
+    Ok(new_bytes)
+}
+
+// `LimitedFuture` wraps a Future with ResourceLimiter, it will automically
+// statistics the cpu time and io bytes consumed by the future, and do async
+// waiting according the configuration of the ResourceLimiter.
+#[pin_project]
+pub struct LimitedFuture<F: Future> {
+    #[pin]
+    f: F,
+    // `pre_delay` and `post_delay` is used to delay this task, at any time, at most one of the two
+    // is valid. A future can only be polled once in one round, so we uses two field here to
+    // workaround this restriction of the rust compiler.
+    #[pin]
+    pre_delay: OptionalFuture<Compat01As03<Delay>>,
+    #[pin]
+    post_delay: OptionalFuture<Compat01As03<Delay>>,
+    resource_limiter: Arc<ResourceLimiter>,
+    res: Poll<F::Output>,
+}
+
+impl<F: Future> LimitedFuture<F> {
+    #[allow(dead_code)]
+    pub fn new(f: F, resource_limiter: Arc<ResourceLimiter>) -> Self {
+        Self {
+            f,
+            pre_delay: None.into(),
+            post_delay: None.into(),
+            resource_limiter,
+            res: Poll::Pending,
+        }
+    }
+}
+
+impl<F: Future> Future for LimitedFuture<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        if !this.post_delay.is_done() {
+            assert!(this.pre_delay.is_done());
+            std::mem::swap(&mut *this.pre_delay, &mut *this.post_delay);
```

Comment:
Is it a standard solution? a little trickly, though I can't figure out another way.

Reply:
Not sure, but I didn't found a better way either.